Feature/detach testcontext base

linkdotnet opened the pull request:
# Current Problem
Right now there is no way to test what happens when we navigate away from a page. Blazor normally disposes  all the components from top to bottom hierarchy wise (parent first, then children, then children of children and so on). We can't simulate this behavior.

# Possible Solution
The draft describes that we extend the `TestContext` to offer a `Detach` method. This will remove everything under the `FragmentContainer` and force another render cycle. This will lead to the wished dispose behavior of Blazor.

## Advantages
The solution is quite simple without invasive changes throughout many objects.

## Drawbacks
We can't detach arbitrary nodes. Personally for me that isn't a requirement and should be part of the components logic and not bUnit. But maybe you have a different angle on this one.

What I didn't check is if the Blazor renderer always works "synchronous"? In the way that, if we have a `IAsyncDisposable` and we "detach" our component, do we really wait until all async Tasks are awaited before we give back control to the caller (so `Detach` returns)?

linkdotnet commented:
Well I thought I take the lead and make this PR ready including documentation and stuff.

linkdotnet commented on src/bunit.core/Rendering/TestRenderer.cs:
```diff
@@ -122,8 +122,14 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
 	{
-		var root = renderedComponents[1] as IRenderedComponentBase<FragmentContainer>;
-		root?.Render();
+		foreach (var component in renderedComponents)
```
First I had this:
```csharp
var root = renderedComponents[1] as IRenderedComponentBase<FragmentContainer>;
```

If you have another way than enumerating through, I am open.

egil replied:
As far as I can tell, the problem here might be that there can be multiple `root` components` rendered into a single test context. And since triggering the removal of the the root components and its children through `Render` call, the `renderedComponents` dictionary gets mutated, I think we first need to do a `renderedComponents.Values.OfType<IRenderedComponentBase<FragmentContainer>>().ToArray()` and then iterate over the result and call Render on each of them.

egil commented:
Thanks Steven. I really appreciate your efforts. 

Im currently suffering from quite severe back pain, so I'll have to let this PR sit for at least a few days before I can properly dive into it. I'll keep you posted.

linkdotnet commented:
> Thanks Steven. I really appreciate your efforts.
> 
> Im currently suffering from quite severe back pain, so I'll have to let this PR sit for at least a few days before I can properly dive into it. I'll keep you posted.

Sure please take your time and get well soon.

linkdotnet commented on src/bunit.core/Rendering/FragmentContainer.cs:
```diff
@@ -19,6 +19,10 @@ public Task SetParametersAsync(ParameterView parameters)
 		{
 			renderHandle.Render(childContent);
 		}
+		else
```
Maybe we should explicitly check here?
```csharp
else if (parameters.ToDictionary().Count == 0)
```
The `operator ==` is not defined for `ParameterView`, therefore the dictionary check.

linkdotnet replied:
> Thank you for your patience. Some feedback finally.
> 
> Added a few code comments, and in general:
> 
> 1. Users can actually test this with the current bUnit. They just need to create a component in their test project, that has a toggle that turns of child content rendering and thus removes child components. Then they just need to render their component under test inside that.
> 2. We need to test this with multiple root components (aka. multiple calls to `Render`/`RenderComponent` during the same test as well as call to `MarkupMatches(renderFragment)`, which also uses the renderer to render html out).
> 3. We need to investigate how components added to the [ctx.RenderTree](https://bunit.dev/docs/providing-input/root-render-tree.html) behave with this.

For 1. Yes that is true. That is more a convenience feature then to minimize the setup needed for such circumstances.

I will write some more tests if you still think this is a valid feature and should be included in the library :D

egil commented on src/bunit.web/TestContext.cs:
```diff
@@ -73,6 +73,14 @@ public virtual IRenderedComponent<TComponent> Render<TComponent>(RenderFragment
+	/// <summary>
+	/// Detaches any <see cref="RenderFragment"/> from the active RenderTree. This forces all children to get disposed
+	/// </summary>
+	public void DetachAllComponents()
```
```suggestion
	public void DetachComponents()
```

I think I like the name `DetachComponents` better. Its more in line with e.g. `FindComponents` and similar functions, where we do not have the `All` word in there.

egil commented on src/bunit.core/Rendering/TestRenderer.cs:
```diff
@@ -117,6 +118,20 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
 
+
+	/// <inheritdoc />
+	public void DetachRoot()
```
Should this not be `DetachRoots` or just `DetachComponents`?

linkdotnet replied:
`DetachComponents` sounds good.
The reasoning behind `DetachRoot` is/was that we disconnect the root element from everthing else, so that from the renderer point of view only the root component exists... but yes your wording does make sense here ;)

egil commented:
Added a few code comments, and in general:

1. Users can actually test this with the current bUnit. They just need to create a component in their test project, that has a toggle that turns of child content rendering and thus removes child components. Then they just need to render their component under test inside that.
2. We need to test this with multiple root components (aka. multiple calls to `Render`/`RenderComponent` during the same test as well as call to `MarkupMatches(renderFragment)`, which also uses the renderer to render html out).
3. We need to investigate how components added to the [ctx.RenderTree](https://bunit.dev/docs/providing-input/root-render-tree.html) behave with this.

linkdotnet commented:
> Thank you for your patience. Some feedback finally.
> 
> Added a few code comments, and in general:
> 
> 1. Users can actually test this with the current bUnit. They just need to create a component in their test project, that has a toggle that turns of child content rendering and thus removes child components. Then they just need to render their component under test inside that.
> 2. We need to test this with multiple root components (aka. multiple calls to `Render`/`RenderComponent` during the same test as well as call to `MarkupMatches(renderFragment)`, which also uses the renderer to render html out).
> 3. We need to investigate how components added to the [ctx.RenderTree](https://bunit.dev/docs/providing-input/root-render-tree.html) behave with this.

For 1. Yes that is true. That is more a convenience feature then to minimize the setup needed for such circumstances.

I will write some more tests if you still think this is a valid feature and should be included in the library :D

egil commented:
> > ... 
> For 1. Yes that is true. That is more a convenience feature then to minimize the setup needed for such circumstances.
> 
> I will write some more tests if you still think this is a valid feature and should be included in the library :D

I like the feature. It will help this particular scenario quite a bit, so please continue by all means.

linkdotnet commented:
For point 2. I added a test and adopted the logic to include a lock. Otherwise we can run into the situation that the same components gets disposed multiple times.

egil commented:
> For point 2. I added a test and adopted the logic to include a lock. Otherwise we can run into the situation that the same components gets disposed multiple times.

How is that possible? A component can only exist one time in the component tree maintained by the Blazor renderer.

linkdotnet commented:
Okay maybe my fault.. seems like it is working without the ? lock` as well. I swear i had the dispose method called multiple times.
Maybe thanks to some weird caching issue by my IDE.
Anyway I removed it

linkdotnet commented:
Regarding `MarkupMatches`: This will throw an `ComponentDisposedException` as it (in my opinion) should.

egil commented:
> Okay maybe my fault.. seems like it is working without the ? lock` as well. I swear i had the dispose method called multiple times. Maybe thanks to some weird caching issue by my IDE. Anyway I removed it

I am not sure. It might be that there is a race condition, or that there are multiple dispose calls happening. Lets debug this and make sure. Find the simplest example and take it from there.

linkdotnet commented:
For your last point (3.) that seems to be a bit tricky as components added via `RenderTree.Add<TComponent>` do not have a `FragmentContainer` as parent and are therefore are not automatically disposed.  have to check later how to tackle this

linkdotnet commented:
> > Regarding `MarkupMatches`: This will throw an `ComponentDisposedException` as it (in my opinion) should.
> 
> Can you share the test example?
```
[Fact]
public void MultipleRenderComponentsShouldBeDisposed()
{
	var cut = RenderComponent<ChildDispose>();

	DetachComponents();

	cut.MarkupMatches("It really doesn't matter what is here");
}
```

egil commented:
> > > Okay maybe my fault.. seems like it is working without the ? lock` as well. I swear i had the dispose method called multiple times. Maybe thanks to some weird caching issue by my IDE. Anyway I removed it
> > 
> > 
> > I am not sure. It might be that there is a race condition, or that there are multiple dispose calls happening. Lets debug this and make sure. Find the simplest example and take it from there.
> 
> haha I could reproduce it right in that moment: ![image](https://user-images.githubusercontent.com/26365461/158174474-26221aa7-f501-4aa4-a965-8d3da8b27842.png)

Try adding a `await Task.Delay(10000000);` at the end of the test and put breakpoints in the dispose method in the CUT. Then for each time the dispose method is hit, investigate the stack to see where the call is coming from. It might be a bug in bUnit.

egil commented:
> > > Regarding `MarkupMatches`: This will throw an `ComponentDisposedException` as it (in my opinion) should.
> > 
> > 
> > Can you share the test example?
> 
> ```
> [Fact]
> public void MultipleRenderComponentsShouldBeDisposed()
> {
> 	var cut = RenderComponent<ChildDispose>();
> 
> 	DetachComponents();
> 
> 	cut.MarkupMatches("It really doesn't matter what is here");
> }
> ```

Ahh yes, accessing `cut` after the component has been disposed/detached should throw. Agreed.

Do test calling `cut.MarkupMatches(renderFragment)` before calling `DetachComponents()`. It should behave the same as if you have multiple calls to `Render` or `RenderComponent`.

egil commented:
> For your last point (3.) that seems to be a bit tricky as components added via `RenderTree.Add<TComponent>` do not have a `FragmentContainer` as parent and are therefore are not automatically disposed. have to check later how to tackle this

That might be what we want, actually. `RenderTree` is ment for dependencies for the CUT, but is not directly under test, and then its not really covered by this scenario. What do you think?

linkdotnet commented:
Oh boy. I am an idiot par excellence. Of course there is no issue.
The issue arises because I have a static List and multiple tests are adding strings to that list on dipose.

Of course there are more entries in that list, when I run multiple test =(

Working with state (static stuff) is just plain dangerous.

linkdotnet commented:
I added some more tests. Please let me know what I can cover additionally.

egil commented on CHANGELOG.md:
```diff
@@ -6,6 +6,9 @@ All notable changes to **bUnit** will be documented in this file. The project ad
 ## [Unreleased]
 
+### Added
+- Added `TestContext.DetachAllComponents` to remove all components from the render tree. By  [@linkdotnet](https://github.com/linkdotnet).
```
```suggestion
- Added `DetachComponents` to `TextContext`, that removes and disposes all components rendered via `Render` or `RenderComponent`. Use this to test component under test's `Dispose` or `DisposeAsync` methods. By  [@linkdotnet](https://github.com/linkdotnet).
```

egil replied:
```suggestion
- Added `DisposeComponents` to `TestContext`. It will dispose and remove all components rendered by via the `TestContext`. By  [@linkdotnet](https://github.com/linkdotnet).
```

egil commented on docs/site/docs/providing-input/root-render-tree.md:
```diff
@@ -31,3 +31,10 @@ Sometimes common test setup logic exists outside the test class, perhaps abstrac
 
 In the listing above, the cascading value `BAR?` is added only if there is not another `CascadingValue<string>` component added to the render tree already.
+
+## Removing all components from the root render tree
```
I am inclined to think this belongs as a sub page under https://bunit.dev/docs/interaction instead of in te render tree.

Also, I think the text needs to reflect that this does not involve components added to `RenderTree`, so the talking about removing components from the root render tree is probably not the way we want to word it.

I think framing it as the ability to remove and dispose explicitly rendered components.

linkdotnet replied:
Sure will do so ;)

egil commented on src/bunit.web/TestContext.cs:
```diff
@@ -73,6 +73,14 @@ public virtual IRenderedComponent<TComponent> Render<TComponent>(RenderFragment
 		=> (IRenderedFragment)this.RenderInsideRenderTree(renderFragment);
 
+	/// <summary>
+	/// Detaches any <see cref="RenderFragment"/> from the active RenderTree. This forces all children to get disposed
```
I dont think it is correct to call it `the active render tree`. Every time you call Render or RenderComponent, a new render tree is created (happens when we call [AssignRootComponentId](https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.core/Rendering/TestRenderer.cs#L218)).

egil commented on src/bunit.core/Rendering/TestRenderer.cs:
```diff
@@ -117,6 +118,17 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
+	/// <inheritdoc />
+	public void DetachComponents()
+	{
+		var rootComponents = renderedComponents.Values.OfType<IRenderedComponentBase<FragmentContainer>>().ToArray();
```
What if we instead look for components of type `WrapperComponent`. That is the type of component that starts a new render tree when a call to Render or RenderComponent happens. Using that should also dispose any components added to `RenderTree`, which makes sense I think.

```suggestion
		var rootComponents = renderedComponents.Values.OfType<WrapperComponent>().ToArray();
```

egil replied:
This is the relevant line where the WrapperComponent gets instantiated:

https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.core/Rendering/TestRenderer.cs#L217

egil replied:
I think it can be made to work by adding a `Detach` method to `WrapperComponent` that calls `renderHandle.Render(static builder => {})`.

egil replied:
Then instead of calling `Render` in `DetachComponents` in `TestRenderer`, we call `Detach`. Thats also much cleaner code.

linkdotnet replied:
I tried this out and we run into the same issues. Mainly because there is no `WrapperComponent` involved when `FindComponent` is called. Therefore the `cut` is not attached to any `WrapperComponent`. Only elements added by the `RenderTree.Add` method.

egil replied:
I am sorry if I am rehashing this, but I am not sure I get why that would not work. If `WrapperComponent` triggers a render that does not have any child components, then ANY previously rendered components should be disposed, independent of they have been returned using `FindComponent` or not. 

Ill try out a few experiments myself to understand this better.

linkdotnet replied:
Please feel free and don't worry. I like to discuss those things... even though via chat is sometimes a bit cumbersome 🗡️ 

>  If WrapperComponent triggers a render that does not have any child components, then ANY previously rendered components should be disposed [...]

That is true. Unfortunately there is no `WrapperComponent` involved anymore (due to `FindComponent`).

![image](https://user-images.githubusercontent.com/26365461/158361424-a6a22730-1f87-4539-859f-e6d44373c41f.png)

egil replied:
Ahh yes, `renderedComponents` doesn't hold the raw components, but `IRenderedComponent`s. Can we use something else then `renderedComponents` to get at the rendered components.

In essence, we basically want the test renderer to dispose all components. A quick look through [Renderer.Dispose(bool)](https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,982) makes it seems we can just call that in `DetachComponents`. Worth a try. Just need to make sure we can perform renders again after that.

linkdotnet replied:
There is an inherit benefit of disposing the `Renderer` itself: It works with elements added via the `RootRenderTree`, but we have to create a new instance of the `TestRenderer` afterwards, which without further changes of the code will not be so easy todo. Especially because we are operating on the `Renderer` from the DI-Container. We have to account for multithreading, ... so at least we would need a lock inside `TestContext.DetachComponents`.

Furthermore I see a logical problem: We call `TestContext.DetachComponents` where the component under test might throw an exception. Now how do we test that?
`this.Renderer.UnhandledException` is not available anymore as we disposed the renderer and created a new one.

Also `WaitFor...` would suffer from similar issues.

egil replied:
> Furthermore I see a logical problem: We call `TestContext.DetachComponents` where the component under test might throw an exception. Now how do we test that?

This is a good edge case, and we need a way to test this. Lets make sure it is possible to catch/verify that an exception is thrown when `DisposeComponents` is called.

linkdotnet commented:
Guess I'll check as well what happens with components added via `ComponentFactory`

egil commented:
It seems there might be a need for WaitForAssertion in these tests, as the they sometimes fail in Linux but not Windows.

linkdotnet commented:
> It seems there might be a need for WaitForAssertion in these tests, as the they sometimes fail in Linux but not Windows.

It was just bad test design :D I fixed that.

egil commented:
> By the way big thanks to you @egil for taking your time and going through the little details and offering alternatives. Really appreciate it!

Not at all. I appreciate your patience and attention to detail. Its really important to me to get things right with bUnit. Once we release code to users its hard to change, to getting API design and implementation details right takes some time.

linkdotnet commented:
I checked out your patch. My already added test, which checks that a component added via `RenderTree.Add` (`RenderFragmentAddedViaRenderTreeShouldNotGetDisposed`) does not fail :D
It can't as elements added via `RenderTree.Add` do net get wrapped in a `WrapperComponent`.

Therefore I'd go with the simpler approach. What do you think?

egil commented on docs/site/docs/interaction/dispose-components.md:
```diff
@@ -0,0 +1,17 @@
+---
+
+# Disposing components
+
+Sometimes it is necessary to remove all components from the root render tree. With this you can simulate the behavior in Blazor when you navigate to a new site and all your components are detached from the render tree. This will dispose all components in the order from parent to children. It is especially helpful if you have event handlers which should be detached to avoid memory leaks.
+
+bUnit comes with a helper method which makes it possible to trigger this behavior.
+
+Just call [`DetachComponents`](xref:Bunit.TestContext.DetachComponents).
```
```suggestion
To dispose all components rendered with a `TestContext`, use the [`DetachComponents`](xref:Bunit.TestContext.DetachComponents) method.  Calling this method will dispose all rendered components, calling any `IDisposable.Dispose` or/and `IAsyncDisposable.DisposeAsync` methods they might have, and remove the components from the render tree, starting with the root components and then walking down the render tree to all the child components.

Disposing rendered components enables testing of logic in `Dispose` and `DisposeAsync` methods, e.g., event handlers, that should be detached to avoid memory leaks.

The following example of this:
```

I like to write my documentation in a terse and non-personal tone, e.g. dont talk directly to the person using words like "you" or "your". According to documentation folks smarter then me, this makes it easier and faster to read.

egil commented on docs/site/docs/interaction/dispose-components.md:
```diff
@@ -0,0 +1,17 @@
+
+[!code-csharp[](../../../samples/tests/xunit/DetachComponentsTest.cs#L10-L15)]
+
+## RenderFragments added via RenderTree
+If you add a `RenderFragment` via the `TestContexts` [`RenderTree.Add`](xref:Bunit.Rendering.RootRenderTree.Add``1(System.Nullable{Action{Bunit.ComponentParameterCollectionBuilder{``0}}})) this will not be disposed via the [`DetachComponents`](xref:Bunit.TestContext.DetachComponents) method.
```
Is this still correct. I think we solved this.

linkdotnet replied:
For short: No we did not. Even with your provided patch, components added directly via the `RootRenderTree` do **not** get disposed via `DisposeComponents`.

egil commented on docs/samples/components/DisposableComponent.razor:
```diff
@@ -0,0 +1,8 @@
+﻿@implements IDisposable
```
Lets make this test a bit more real world. You have a great example in the docs with event handlers detachment. Lets write sample that shows this.

linkdotnet replied:
I went with `CancellationToken` as this was more compact (and one of the reasons I created that PR :D)

egil replied:
I get that, but I think my NavigationManager example is more simple, and doesnt require us introducting additional types that people need to understand when reading the docs.

egil commented on docs/site/docs/providing-input/root-render-tree.md:
```diff
@@ -30,4 +30,4 @@ Sometimes common test setup logic exists outside the test class, perhaps abstrac
 [!code-csharp[](../../../samples/tests/xunit/RenderTreeTest.cs#L36-L38)]
 
-In the listing above, the cascading value `BAR?` is added only if there is not another `CascadingValue<string>` component added to the render tree already.
+In the listing above, the cascading value `BAR?` is added only if there is not another `CascadingValue<string>` component added to the render tree already.
```
Hmm I cannot see any changes in this file? is it whitespace or line endings that have changed?

linkdotnet replied:
Me neither :D but I'll try to revert that :D

egil commented on src/bunit.core/Rendering/ITestRenderer.cs:
```diff
@@ -73,4 +73,9 @@ IRenderedComponentBase<TComponent> FindComponent<TComponent>(IRenderedFragmentBa
+	/// <summary>
+	/// Detaches the root component from the any other component and enters another render cycle
+	/// </summary>
+	public void DetachComponents();
```
I know we already talked about the naming of this method, but I feel inspired to change it to `DisposeComponents`, because the word "detach" implies that the components can be reattached again, which is not the case. `DisposeComponents` aligns closely with what actually happens.

linkdotnet replied:
That's fine for me. Will rename

egil commented on src/bunit.core/Rendering/ITestRenderer.cs:
```diff
@@ -73,4 +73,9 @@ IRenderedComponentBase<TComponent> FindComponent<TComponent>(IRenderedFragmentBa
 	IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TComponent>(IRenderedFragmentBase parentComponent)
 		where TComponent : IComponent;
+
+	/// <summary>
+	/// Detaches the root component from the any other component and enters another render cycle
+	/// </summary>
+	public void DetachComponents();
```
```suggestion
	/// <summary>
	/// Disposes the all component rendered by the <see cref="ITestRenderer" />.
	/// </summary>
	public void DisposeComponents();
```

egil commented on src/bunit.core/Rendering/TestRenderer.cs:
```diff
@@ -117,6 +118,17 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
 
+
+	/// <inheritdoc />
+	public void DetachComponents()
```
```suggestion
	public void DisposeComponents()
```

egil commented on src/bunit.web/TestContext.cs:
```diff
@@ -73,6 +73,14 @@ public virtual IRenderedComponent<TComponent> Render<TComponent>(RenderFragment
 	public virtual IRenderedFragment Render(RenderFragment renderFragment)
 		=> (IRenderedFragment)this.RenderInsideRenderTree(renderFragment);
 
+	/// <summary>
+	/// Detaches any <see cref="RenderFragment"/> from the RenderTree. This forces all children to get disposed
+	/// </summary>
+	public void DetachComponents()
+	{
+		Renderer.DetachComponents();
+	}
+
```
```suggestion
	/// <summary>
	/// Disposes all components rendered via this <see cref="TestContext"/>.
	/// </summary>
	public void DisposeComponents()
	{
		Renderer.DisposeComponents();
	}

```

egil reviewed (commented):
I think I have changed my mind on calling the method DetachComponnets. I think DisposeComponents are better. See comments above on details.

A few notes:

1. I think we need to have a test with `IAsyncDisposable` as well. It behaves differently, its async, so lets create a test component that does something async like calling `await Task.Delay(10)`.
2. We need a test that shows we can catch exceptions thrown by a Dispose/DisposeAsync method in a component and assert on that.
3. It would be good to add a test that shows content in a `RenderFragment` gets disposed correctly as well, when rendered via the `TestContext.Render(renderFragment)` method. I think it will just work, but its a good test to have in case we refactor later.
4. I currently prefer the patch version I shared where the renderer keeps track of the `WrapperComponent`s and use these to dispose and clean up. It seems simpler to me to be honest, and properly cleans everything up at the top level. `WrapperComponent` is the root component everything is rendered inside, so I am also considering renaming it to `RootComponent`, just to make that more obvious.

egil commented on docs/samples/components/DisposableComponent.razor:
```diff
@@ -1,8 +1,18 @@
+    _cts.Cancel();
+    _cts.Dispose();
   }
 }
```
I think we can simplify this quite a bit, something like this perhaps:

```razor
@implements IDisposable
@inject NavigationManager NavigationManager
@code {
  protected override void OnInitialized()
  {
    NavigationManager.LocationChanged += OnLocationChanged;
  }

  public void Dispose()
  {
    NavigationManager.LocationChanged -= OnLocationChanged;
  }
  
  private void LocationChanged(object sender, LocationChangedEventArgs e)
  {
    // ...
  }
}
```

egil commented on src/bunit.core/Rendering/ITestRenderer.cs:
```diff
@@ -75,7 +75,7 @@ IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TComponent>(IRe
 
 	/// <summary>
-	/// Detaches the root component from the any other component and enters another render cycle
+	/// Disposes the all components
```
```suggestion
	/// Disposes all components rendered by the <see cref="ITestRenderer" />.
```

egil commented on tests/bunit.web.tests/BlazorE2E/DisposeComponentsTest.cs:
```diff
@@ -0,0 +1,135 @@
+
+namespace Bunit.BlazorE2E;
+
+public class DisposeComponentsTest : TestContext
```
These tests actually belong in the TestContextTest.cs class.

egil commented on tests/bunit.web.tests/BlazorE2E/DisposeComponentsTest.cs:
```diff
@@ -0,0 +1,135 @@
+	{
+		RenderComponent<ThrowExceptionComponent>();
+
+		DisposeComponents();
```
I was actually expecting the call to `DisposeComponents` to throw. That is way more discoverable for the end user then having to inspect `Renderer.UnhandledException`. See comment on Renderer.

egil commented on src/bunit.core/Rendering/TestRenderer.cs:
```diff
@@ -120,13 +120,17 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
+			}
+		});
+
+		rootComponents.Clear();
```
Lets write it something like this, then any unhandled exceptions are throw back to the user immidiately:

```c#
		var disposeTask = Dispatcher.InvokeAsync(() =>
		{
			foreach (var root in rootComponents)
			{
				root.Detach();
			}
		});

		if (!disposeTask.IsCompleted)
		{
			disposeTask.GetAwaiter().GetResult();
		}

		rootComponents.Clear();
		AssertNoUnhandledExceptions();
```

linkdotnet replied:
I just checked this piece of code. Quite funny: That works for `IDisposable` but not for `IAsyncDIsposable`. Even though `DisposeAsync` wasn't even called the `disposeTask.IsCompleted` is `true`.

Therefore we have the two ways of handling exceptions:
 * `IDisposable` -> `DisposeComponents` would directly throw an exception
 * `IAsyncDisposable` -> `DisposeComponents` does not throw directly because the underlying `Task` is completed for some reason (even though `DisposeAsync` is not done). We would have to take the: `var exception = Renderer.UnhandledException.Result` approach (or `var exception = await Renderer.UnhandledException`).

linkdotnet replied:
If you want to reproduce this behavior execute `ExceptionThrownInDisposeAsyncShouldBeCatchable`

egil replied:
> * `IAsyncDisposable` -> `DisposeComponents` does not throw directly because the underlying `Task` is completed for some reason (even though `DisposeAsync` is not done). We would have to take the: `var exception = Renderer.UnhandledException.Result` approach (or `var exception = await Renderer.UnhandledException`).

That is interesting. Let dig into Blazors renderer and see why that is.

egil replied:
So it looks like there is nothing we can do about async disposers throwing (dispose should never throw anyway): https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,829

linkdotnet replied:
> So it looks like there is nothing we can do about async disposers throwing (dispose should never throw anyway): https://source.dot.net/#Microsoft.AspNetCore.Components/RenderTree/Renderer.cs,829

Ah thanks for the link. I guess we leave it as it is, also documentation wise if someone wants to check for exceptions.

egil reviewed (commented):
Thanks for the update. Your changes inspired some more suggestions. Hang on in there, we're close!

egil commented on docs/samples/components/DisposableComponent.razor:
```diff
@@ -0,0 +1,20 @@
+﻿@using Microsoft.AspNetCore.Components.Routing
+@implements IDisposable
+@inject NavigationManager _navigationManager
+@inject ITrackingService _trackingService
+@code {
+  protected override void OnInitialized()
+  {
+    _navigationManager.LocationChanged += OnLocationChanged;
+  }
+
+  public void Dispose()
+  {
+    _navigationManager.LocationChanged -= OnLocationChanged;
+  }
+
+  private void OnLocationChanged(object sender, LocationChangedEventArgs e)
+  {
+    _trackingService.SetTouchPoint(e.Location);
+  }
+}
```
```suggestion
﻿@using Microsoft.AspNetCore.Components.Routing
@implements IDisposable
@inject NavigationManager NavigationManager
@code {
  [Parameter] public Action<string> LocationChangedCallback { get; set; }

  protected override void OnInitialized()
  {
    NavigationManager.LocationChanged += OnLocationChanged;
  }

  public void Dispose()
  {
    NavigationManager.LocationChanged -= OnLocationChanged;
  }

  private void OnLocationChanged(object sender, LocationChangedEventArgs e)
  {
    LocationChangedCallback(e.Location);
  }
}
```

I think we can simplify the demo test by just having a param we pass.

egil commented on docs/samples/tests/xunit/DisposeComponentsTest.cs:
```diff
@@ -0,0 +1,48 @@
+  [Fact]
+  public void DisposeElements()
+  {
+    using var ctx = new TestContext();
+    var trackingService = new Mock<ITrackingService>();
+    ctx.Services.AddScoped<ITrackingService>(_ => trackingService.Object);
+    var cut = ctx.RenderComponent<DisposableComponent>();
+    ctx.DisposeComponents();
+
+    ctx.Services.GetRequiredService<NavigationManager>().NavigateTo("newurl");
+
+    trackingService.Verify(t=> t.SetTouchPoint(It.IsAny<string>()), Times.Never);
```
```suggestion
    using var ctx = new TestContext();
    var calledTimes = 0;
    var cut = ctx.RenderComponent<DisposableComponent>(parameters => parameters
      .Add(p => p.LocationChangedCallback, url => calledTimes++)
    );
    
    ctx.DisposeComponents();

    ctx.Services.GetRequiredService<NavigationManager>().NavigateTo("newurl");

    Assert.Equal(0, calledTimes);
```

What do you think about this?

linkdotnet commented:
I'd merged your suggested changes!Much appreciated the simpler setup. I would also like to squash all commits due to the big commit message otherwise. You squash PR's anyway, not?

egil commented:
Sounds great. I will normally squash a, since it makes a rollback simpler in most cases.

linkdotnet commented:
> Sounds great. I will normally squash a, since it makes a rollback simpler in most cases.

Integrated your changes and squashed the whole thing.
I guess the last open point is `DisposeAsync` which isn't finished even though the renderer indicates it?

egil commented:
Hey Steven, I did a deep dive on the changes live on stream today at twitch.tv/egilhansen and discovered af few things. I commited the changes to https://github.com/bUnit-dev/bUnit/commits/feature/detach-testcontext-base (could not figure out how to push to your branch/the PR), but if you can merge my changes into the PR, I think we are ready to complete it.

There are basically things I did:

1. Move `DisposeComponents` up to `TestContextBase`.
2. Added `DisposeComponents` to `TestContextWrapper`
3. Moved the tests to the correct test class and aligned the style of them (using display name attribute)
4. Updated some of the test components to not use static fields/properties. That will cause tests to break randomly when the test suite is run in parallel. Switch to using cascading parameter to pass in the CallStack variable.

Overall, I am super happy with the solution we have now. Thanks so much for your efforts on this!

linkdotnet commented:
I am currently checking why this one test is faling on the linux box. Locally it works fine :D

linkdotnet commented:
> I am currently checking why this one test is faling on the linux box. Locally it works fine :D

Yeah not sure what is going on. The test was successful multiple times.

linkdotnet commented:
By the way very nice stream. I just watched it to check your changes ;)

To your last question regarding pushing changes to my repository. That should be possible. But I think only with extra work:

Basically you have to add my bUnit repository as another remote:
`git remote add downstream https://github.com/linkdotnet/bUnit.git` 
Then you could in theory work on that branch as you have the permission. Or just merge your changes into my feature branch as well.

Is kind of the revert situation I am in:
My `origin` points to `linkdotnet/bUnit` and `upstream` points to `bUnit/bUnit`.
So I added your branch via
`git fetch upstream && git merge upstream/feature/detach`

egil commented:
Yeah, that async difference between Linux and Windows sucks. The `Test202` test can be written like this, it might fix the issue:

```c#
	[Fact(DisplayName = "DisposeComponents calls DisposeAsync on rendered components")]
	public void Test202()
	{
		var callStack = new List<string>();
		var cut = RenderComponent<AsyncDisposableComponent>(ps => ps.Add(p => p.CallStack, callStack));

		DisposeComponents();

		cut.WaitForAssertion(() =>
			callStack
				.ShouldHaveSingleItem()
				.ShouldBe(nameof(AsyncDisposableComponent)));
	}

	private sealed class AsyncDisposableComponent : ComponentBase, IAsyncDisposable
	{
		[Parameter]
		public ICollection<string> CallStack { get; set; } = Array.Empty<string>();

		public async ValueTask DisposeAsync()
		{
			await Task.Delay(10);
			CallStack.Add(nameof(AsyncDisposableComponent));
		}
	}
```

egil commented:
And to avoid those pesky boolean values, we can do the same for `Test203`:

```c#
	[Fact(DisplayName = "DisposeComponents should dispose components added via ComponentFactory")]
	public void Test203()
	{
		var callStack = new List<string>();
		ComponentFactories.Add<ChildDispose, MyChildDisposeStub>();
		RenderComponent<ParentDispose>(ps => ps.Add(p => p.CallStack, callStack));

		DisposeComponents();

		callStack
			.ShouldHaveSingleItem()
			.ShouldBe(nameof(MyChildDisposeStub));
	}

	private sealed class MyChildDisposeStub : ComponentBase, IDisposable
	{
		[Parameter]
		public ICollection<string> CallStack { get; set; } = Array.Empty<string>();

		public void Dispose()
		{
			CallStack.Add(nameof(MyChildDisposeStub));
		}
	}
```

linkdotnet commented:
> Yeah, that async difference between Linux and Windows sucks. The `Test202` test can be written like this, it might fix the issue:
I quickly tried your suggested solution but the issue still persists: I could boil it down a bit. If those 4 test suites are running together I can make it consistently  failing

![image](https://user-images.githubusercontent.com/26365461/160598284-c740fe78-08aa-4c81-8ada-6f59b439d11a.png)

I am not sure right now, have to dig in deeper later, but my educated guess is, that somehow the `WaitForHelpers` get screwed up big time. Also the test-order is important. When the `Extensions.WaitForHelpers` run prior to our dispose tests it will fail, every single time. Only the other way around the dispose tests get green.

In the next tests i tried only to run the .net5 suite for the `WaitForHelpersTests` but not the .net6. And voilà only the .net5 version of our DisposeAsync is failing:

![image](https://user-images.githubusercontent.com/26365461/160598489-dd17058a-dc7e-4c2d-b70d-b2c14a152e13.png)

So I digged deeper and checked which test(s) causes the issue: And it boils down to this one (there could be more, but this candidate made it stable failign for me):
![image](https://user-images.githubusercontent.com/26365461/160599539-6c9888dd-00f7-4c7c-a111-2464ce55f4e6.png)

I'll let you know if you find out more.

egil commented:
Some quick feedback on your research:

I don't think it's a bug in the renderer. They simply don't have an interest in calling [`UpdateDisplayAsync`](https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.core/Rendering/TestRenderer.cs#L151) after an async disposable component completes it's work, since there would never be a need to update the render tree. However, the `UpdateDisplayAsync` is our hook into notifying "WaitForXxx....", and that never happens. At least, that's my guess at the moment.

So that leads me to think there is no way for us to really use the wait for logic in bUnit in this particular case, and that users will have to fall back to having Task.Delay in their test code if they have async disposable code.

linkdotnet commented:
> Some quick feedback on your research:
> 
> I don't think it's a bug in the renderer. They simply don't have an interest in calling [`UpdateDisplayAsync`](https://github.com/bUnit-dev/bUnit/blob/main/src/bunit.core/Rendering/TestRenderer.cs#L151) after an async disposable component completes it's work, since there would never be a need to update the render tree. However, the `UpdateDisplayAsync` is our hook into notifying "WaitForXxx....", and that never happens. At least, that's my guess at the moment.
> 
> So that leads me to think there is no way for us to really use the wait for logic in bUnit in this particular case, and that users will have to fall back to having Task.Delay in their test code if they have async disposable code.

Fair enough :D still I let the issue open to clarify if this is intended. Arguably the component is not disposed but in disposing. Maybe this is more an implementation detail and from the outside world the observed behavior would be the same.

Anyway I updated the PR and removed the `WaitForAssertion` part and added a small note in the docs about that behavior.
If the asp.net team still decides to make it observable for us with still can introduce `WaitForDisposed` or similar.

egil commented:
Second point on your investigation: the reason why `Task.Delay(10)` sometimes works on some machines (windows vs. linux, intel vs. amd) has to do with the schedular in the OS and the timing of it, as far as I know. I remember somebody mention that if we had set it to something >20, then we might not have seen a failing test always. It has to do with how long the schedular will wait on an async operation before it is declared "really" async, and causes the executing thread to branch off an go back to doing other things, i.e. spin up an async context that will continue on another thread.

linkdotnet commented:
> Second point on your investigation: the reason why `Task.Delay(10)` sometimes works on some machines (windows vs. linux, intel vs. amd) has to do with the schedular in the OS and the timing of it, as far as I know. I remember somebody mention that if we had set it to something >20, then we might not have seen a failing test always. It has to do with how long the schedular will wait on an async operation before it is declared "really" async, and causes the executing thread to branch off an go back to doing other things, i.e. spin up an async context that will continue on another thread.

Haha yes :D see my last commit ;)

egil commented on src/bunit.core/Rendering/FragmentContainer.cs:
```diff
@@ -19,6 +19,10 @@ public Task SetParametersAsync(ParameterView parameters)
 		{
 			renderHandle.Render(childContent);
 		}
+		else
+		{
+			renderHandle.Render(_ => {});
+		}
```
I think we can roll back this change, right?

linkdotnet replied:
Good catch. No this is leftover from the "old" approach

egil commented on src/bunit.core/Rendering/TestRenderer.cs:
```diff
@@ -117,6 +118,27 @@ public IReadOnlyList<IRenderedComponentBase<TComponent>> FindComponents<TCompone
+	/// <inheritdoc />
+	public void DisposeComponents()
+	{
+		var disposeTask = Dispatcher.InvokeAsync(() =>
+		{
+			foreach (var root in rootComponents)
+			{
+				root.Detach();
+			}
+		});
+
+		if (!disposeTask.IsCompleted)
+		{
+			disposeTask.GetAwaiter().GetResult();
+		}
```
```suggestion
		// The dispatcher will always return a completed task, 
		// since calling root.Detach does not trigger any (async) Dispose
		// operations directly on the components. It happens
		// indirectly after a re-render, where the components
		// are removed from the render tree.
		Dispatcher.InvokeAsync(() =>
		{
			foreach (var root in rootComponents)
			{
				root.Detach();
			}
		});		
```

Lets simplify this and leave an explanation for our future selves :)

linkdotnet replied:
Well the explanation is not completely true and I would like not to delete the `diposeTas.GetAwaiter().GetResult()` stuff as we need this for throwing an expcetion in `TestContextBase.DisposeComponents` if our component under test will itself throw the exception in `Dispose`.

The sync version will lead to a direct update of the renderer...only `DisposeAsync` is special

linkdotnet replied:
Though I will add the comment for the async version for our future self :D

egil reviewed (commented):
Lets rename WrapperComponent to RootComponent, now that we are neck deep in this code.

Otherwise here are some final comments. Then I believe we are done.

linkdotnet commented:
By the way that "EndOfStream" exceptions thanks to the collector...very annoying. I did not found any workaround for that. Right now I am even thinking of disabling test coverage collection for my projects due to this issue.

linkdotnet commented:
Yeeaahhhh 🎉🎉🎉✨✨✨
Much appreciated @egil on this one! It started so innocent but mainly thanks to you, learnt a lot how Blazor operates!

egil commented:
Thank you for the hard work in this. And yes, as soon as you start touching the renderer things tend to get a bit more  complex.
